### DISCUSSION TITLE
How to prevent spaces being changed into '%20' when using file links?

### DISCUSSION THREAD

**Discussion by Robert-Oxford** [quarto-dev/quarto-cli#10231] [Q&A]
### Description

I want to insert a hyperlink to a file I have stored on my computer. I am using Windows 10.

When I include a link as below, the exported versions in both HTML and pdf subsitute %20 for all of the spaces. This would work on a web-based URL .. but does not work on windows as a file reference.

```
''''qmd
---
title: "Dealing with spaces in file hyperlinks on windows"
author: "Rob"
date: "4th July 2024"

format:
  html:
    toc: true
    number-sections: true
    css: styles.css
---


[My file](file:///C:\Users\Fred Bloggs\Doc 000-000-0001.docx)
''''
```
Produces this:

`<p><a href="file:///C:\Users\Fred%20Bloggs\Doc%20000-000-0001.docx">My file</a></p>`

Which does not work on windows (I get a 'file not found error")

I need this:

`<p><a` href="file:///C:\Users\Fred Bloggs\Doc 000-000-0001.docx">My file</a></p>

**Comment by mcanouil** [comment]
I am not sure you can or maybe using the raw syntax, see <https://quarto.org/docs/authoring/markdown-basics.html#raw-content>.
(I can't test as I don't have access to Quarto on a Windows machine)

Note that creating links (independently of the protocol) with spaces in it is really considered as a bad practice for obvious reasons such as the issue you are facing.

**Comment by Robert-Oxford** [comment]
Thank you - I will certainly try the 'raw' option. I agree about the spaces in links. In this case I need to create links into a repository of existing documents and files. Those documents have been created by many people who use Windows as their primary operating system and are used to using spaces in file names. As you say - it is not desirable, but at the moment I am stuck with it. I may, in the future, do a bulk renaming of documents in the repository but as you can imagine, this can easily cause problems in itself. Thank you for your help.

**Comment by cderv** [comment]
I believe this is due to own Pandoc behavior that will HTML escaped the target part of the link 
````
> quarto pandoc --to html
[My file](file:///C:\Users\Fred Bloggs\Doc 000-000-0001.docx)
^Z
<p><a href="file:///C:\Users\Fred%20Bloggs\Doc%20000-000-0001.docx">My
file</a></p>
````

Pandoc does URL escape in the Reader directly 
````powershell
> quarto pandoc --to native
[My file](file:///C:\Users\Fred Bloggs\Doc 000-000-0001.docx)
^Z
[ Para
    [ Link
        ( "" , [] , [] )
        [ Str "My" , Space , Str "file" ]
        ( "file:///C:\\Users\\Fred%20Bloggs\\Doc%20000-000-0001.docx"
        , ""
        )
    ]
]
````

So as this is quite specific right now, I do think you should leverage a [custom filter](https://quarto.org/docs/extensions/filters.html) option to catch those links of yours, and unnescape what needs to be 

For example `custom.lua`
````lua
Link = function(l)
  if (l.target:match('^file://')) then
    l.target = l.target:gsub("%%20", " ")
  end
  return l
end
````

and then in your document or project 
````yaml
filters:
  - custom.lua
````

You could even add an attribute or class to catch on the link to decode those only. 

You may need more than `%20` characters. So maybe there is more replacement to do. I don't know if a file:: url needs to be fully decoded or not. 

FWIW we have internal function for this kind of things but we don't export them yet. 
https://github.com/quarto-dev/quarto-cli/blob/913292ce3606f454090f406e5d89894ca3a18b31/src/resources/filters/common/url.lua#L4-L24 

Hope it helps

**Reply by Robert-Oxford** [reply]
Thank you so much. I really do appreciate such a thoughtfull response. I will give it a try.
